routing/dht/routing.go

```diff
@@ -111,20 +116,27 @@ func (dht *IpfsDHT) GetValue(ctx context.Context, key u.Key) ([]byte, error) {
 // Provide makes this node announce that it can provide a value for the given key
 func (dht *IpfsDHT) Provide(ctx context.Context, key u.Key) error {
 
+	log.Event(ctx, "provideBegin", &key)
+	defer log.Event(ctx, "provideEnd", &key)
 	dht.providers.AddProvider(key, dht.self)
-	peers := dht.routingTable.NearestPeers(kb.ConvertKey(key), PoolSize)
-	if len(peers) == 0 {
-		return nil
+
+	peers, err := dht.getClosestPeers(ctx, key)
+	if err != nil {
+		return err
 	}
 
-	//TODO FIX: this doesn't work! it needs to be sent to the actual nearest peers.
-	// `peers` are the closest peers we have, not the ones that should get the value.
-	for _, p := range peers {
-		err := dht.putProvider(ctx, p, string(key))
-		if err != nil {
-			return err
-		}
+	wg := sync.WaitGroup{}
```

Comment:
see https://github.com/jbenet/go-ipfs/pull/456/files#r22405941 -- this should be a query too